src/main/scala/com/fulcrumgenomics/bam/Bams.scala

```diff
@@ -0,0 +1,207 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2017 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.bam
+
+import com.fulcrumgenomics.FgBioDef._
+import com.fulcrumgenomics.util.{BetterBufferedIterator, ProgressLogger}
+import dagr.commons.util.LazyLogging
+import htsjdk.samtools.SAMFileHeader.{GroupOrder, SortOrder}
+import htsjdk.samtools.util.{SequenceUtil, SortingCollection}
+import htsjdk.samtools._
+import htsjdk.samtools.reference.ReferenceSequenceFileWalker
+
+import scala.collection.JavaConversions.iterableAsScalaIterable
+import scala.collection.mutable.ListBuffer
+
+/**
+  * Class that represents all reads from a template within a BAM file.
+  */
+case class Template(r1: Option[SAMRecord],
+                    r2: Option[SAMRecord],
+                    r1Supplementals: Seq[SAMRecord] = Nil,
+                    r2Supplementals: Seq[SAMRecord] = Nil,
+                    r1Secondaries:   Seq[SAMRecord] = Nil,
+                    r2Secondaries:   Seq[SAMRecord] = Nil) {
+
+  /** Gets the query name of the template. */
+  val name: String = {
+    if (r1.isDefined) r1.get.getReadName
+    else if (r2.isDefined) r2.get.getReadName
+    else Seq(r1Supplementals, r2Supplementals, r1Secondaries, r2Secondaries).find(_.nonEmpty).map(_.head.getReadName)
+      .getOrElse(throw new IllegalStateException("Template created with no reads!"))
+  }
+
+  /** Returns an iterator over all records that are not the primary r1 and r2. */
+  def allSupplementaryAndSecondary: Iterator[SAMRecord] =
+    r1Supplementals.iterator ++ r2Supplementals.iterator ++ r1Secondaries.iterator ++ r2Secondaries.iterator
+
+  /** The total count of records for the template. */
+  def readCount: Int = r1.size + r2.size + r1Supplementals.size + r2Supplementals.size + r1Secondaries.size + r2Secondaries.size
```

Comment:
It's currently a `def` not a `val`, but making it a `lazy val` is a good idea anyway.